### DISCUSSION TITLE
Re testing.

### DISCUSSION THREAD

**Discussion by fade2gray** [mark-wiemer/ahkpp#37] [General]
With regard to my 'limited test' [here](https://github.com/mark-wiemer/vscode-autohotkey-plus-plus/pull/32) and though I have Git for Windows desktop installed, my understanding of Git is very limited.

**Comment by FuPeiJiang** [comment]
by `Git for Windows desktop` do you mean `GitHub Desktop`?, to be fair, `GitHub Desktop` is very limited: only uses a few git commands

for Git GUI, personally, I use [smartgit](https://www.syntevo.com/smartgit/): the WinTitle is the current git repo I am in. so I can easily use that in ahk: run commands using hotkeys, on the git repo I am currently in

first, learn the command line, do you know cd: change directory ? cd to your git repo

then, learn git, learn to use the commands, then you can use abstractions, alias, to simplify tasks.
learn by doing, create a git repo for learning, to run commands in that repo without fear.
so your first exercise should be to `checkout an upstream branch.`

cd to your local fork of this repo, run `git remote`, tell me what you got.
it should be
```
origin
upstream
```
origin is your own (fade2gray/vscode-autohotkey-plus-plus)
upstream is mark-wiemer/vscode-autohotkey-plus-plus

`upstream/fix-18` would be branch `fix-18` from mark-wiemer/vscode-autohotkey-plus-plus

now, `fix-18` from mark-wiemer/vscode-autohotkey-plus-plus has been deleted, so do not try to get it
___
what is `git checkout`:
a git repo has many branches, only one branch can be the "active one", `git checkout <branch>` sets `<branch>` as the active branch:
your folder will be rewritten/synced to the files in HEAD of `<branch>`

this assumes `<branch>` exists locally, if it doesn't, ... 
this is your exercise, good luck searching
"how to checkout a branch that doesn't exist locally"

<details>
  <summary>Answer was given here</summary>

https://github.com/mark-wiemer/vscode-autohotkey-plus-plus/pull/32#issuecomment-770263559  
</details>
<br>
__<br>
if you did not find how, if you want more exercises, ask me<br>
if you can make this explanation clearer please tell me.

**Comment by mark-wiemer** [comment]
Another great git resource is the official [git Reference](https://git-scm.com/docs). I personally don't use any git GUI, I find understanding the command line allows for greater flexibility. But to each their own :)

**Comment by fade2gray** [comment]
@FuPeiJiang 
The only way I can get the `origin / upstream` output when using `git remote` at the command line, is by creating a local `clone` using `GitHub Desktop`.

If I create a clone by using `git clone <repository>` or by using SmartGui or GitKraken, the only output I see when using `git remote` is `origin`.

@mark-wiemer 
Re https://github.com/mark-wiemer/vscode-autohotkey-plus-plus/pull/32#issuecomment-770264089 - I assume "pull this branch" means create a local clone?

If I open the clone in VSCode and press F5 `Start Debugging`, I get the error...

> The terminal process "C:\WINDOWS\System32\WindowsPowerShell\v1.0\powershell.exe -Command npm run watch" terminated with exit code: 1.

**Reply by FuPeiJiang** [reply]
> git clone \<repository\>

what \<repository\> did you use ? please give me the link.

it should be `git clone https://github.com/fade2gray/vscode-autohotkey-pp2.git` instead of `git clone https://github.com/mark-wiemer/vscode-autohotkey-plus-plus.git`

clone your own fork.
___
> I assume "pull this branch" means create a local clone?

you cannot pull that branch, it has been deleted.
pull this branch means download this branch, you will then have that branch locally. (you can rename it, it's like save-as), you cannot rename the branch directly AFTER you clone it).
___
you should learn the vocabulary, by trying in a "test repo"

**Reply by fade2gray** [reply]
I was cloning my own fork, but I didn't realise I had to tick this checkbox in SmartGui for which there is no requirement in Github Desktop.

![Capture](https://user-images.githubusercontent.com/1332009/107776621-b6100a00-6d39-11eb-82c1-de08319b8119.PNG)

**Reply by fade2gray** [reply]
With the `git clone` and `git remote` commands I tried...

```
C:\Users\fade2gray\GitHub>git clone --recurse-submodules https://github.com/fade2gray/vscode-autohotkey-pp2.git vscode-autohotkey-pp2_gc-recsub
Cloning into 'vscode-autohotkey-pp2_gc-recsub'...
remote: Enumerating objects: 284, done.
remote: Counting objects: 100% (284/284), done.
remote: Compressing objects: 100% (180/180), done.
remote: Total 2085 (delta 130), reused 189 (delta 92), pack-reused 1801
Receiving objects: 100% (2085/2085), 958.67 KiB | 2.18 MiB/s, done.
Resolving deltas: 100% (1205/1205), done.

C:\Users\fade2gray\GitHub>git clone --remote-submodules https://github.com/fade2gray/vscode-autohotkey-pp2.git vscode-autohotkey-pp2_gc-remsub
Cloning into 'vscode-autohotkey-pp2_gc-remsub'...
remote: Enumerating objects: 284, done.
remote: Counting objects: 100% (284/284), done.
remote: Compressing objects: 100% (180/180), done.
remote: Total 2085 (delta 130), reused 189 (delta 92), pack-reused 1801
Receiving objects: 100% (2085/2085), 958.67 KiB | 2.23 MiB/s, done.
Resolving deltas: 100% (1205/1205), done.

C:\Users\fade2gray\GitHub>cd vscode-autohotkey-pp2_gc-recsub

C:\Users\fade2gray\GitHub\vscode-autohotkey-pp2_gc-recsub>git remote
origin

C:\Users\fade2gray\GitHub\vscode-autohotkey-pp2_gc-recsub>cd ..

C:\Users\fade2gray\GitHub>cd vscode-autohotkey-pp2_gc-remsub

C:\Users\fade2gray\GitHub\vscode-autohotkey-pp2_gc-remsub>git remote
origin
```

**Reply by FuPeiJiang** [reply]
I don't think you need `--recurse-submodules`, if you don't know what it does, you probably don't need it.
___

if you don't have the remote `upstream` add it using: `git remote add upstream https://github.com/ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git`
in your case: `git remote add upstream https://github.com/mark-wiemer/vscode-autohotkey-plus-plus.git`

`upstream` is just a name for a remote: you can name it anything

for example : `git remote add mark-wiemer https://github.com/mark-wiemer/vscode-autohotkey-plus-plus.git`

**Reply by mark-wiemer** [reply]
To confirm, `--recurse-submodules` is not necessary. It doesn't hurt, but this project has no submodules (and I don't anticipate it ever will have submodules). You can learn about submodules at [the official submodule docs](https://git-scm.com/book/en/v2/Git-Tools-Submodules).
